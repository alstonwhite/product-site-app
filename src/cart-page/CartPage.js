import React from 'react';

import CartTile from './CartTile'
import { useSelector } from "react-redux";

const CartPage = ({onUpdate, onRemove}) => {

    const cart = useSelector(state => state.cart); 
    let cartValue = cart ? parseInt(cart.reduce((sum, cur) => sum + cur.price * cur.quantity, 0)) : 0;

    return (
        <div className="cart-page" id="cart-page">
            <h1>Cart</h1>
            <div className="cart-grid__items">
                {cart.map(product => (
                    <CartTile 
                        product={product}
                        onUpdate={onUpdate}
                        onRemove={onRemove}
                    />
                ))}
            </div>
            <div className="cart-page__footer">
                <div className="cart-page__footer__value">
                    {`Total: $${cartValue}`}
                </div>
                <a 
                    href='#' 
                    className ="button cart-page__footer__checkout"
                    // onClick={() => checkout()}
                >
                    Checkout
                </a>
                {/* change to button? */}
            </div>
        </div>
    );
  };
    
    export default CartPage;