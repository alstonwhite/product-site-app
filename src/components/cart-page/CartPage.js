import React, {useState, useEffect} from 'react';

import CartTile from './CartTile'
import { useSelector } from "react-redux";
import { stripeToken, checkout } from '../../utils/checkoutStripe'

const CartPage = () => {

    const [stripe, setStripe] = useState(null)

    const cart = useSelector(state => state.cart); 
    let cartValue = cart ? parseInt(cart.reduce((sum, cur) => sum + cur.price * cur.quantity, 0)) : 0;

    useEffect(() => {
        if (window.Stripe) setStripe(window.Stripe(stripeToken))
    }, [])

    return (
        <div className="cart-page" id="cart-page">
            <h1>Cart</h1>
            <div className="cart-grid__items">
                {cart.map(product => (
                    <CartTile 
                        product={product}
                        key={product.id}
                    />
                ))}
            </div>
            <div className="cart-page__footer">
                <div className="cart-page__footer__value">
                    {`Total: $${cartValue}`}
                </div>
                <button
                    className ="cart-page__footer__checkout btn"
                    onClick={() => checkout(stripe, cart)}
                >
                    Checkout
                </button>
            </div>
        </div>
    );
};
    
export default CartPage;