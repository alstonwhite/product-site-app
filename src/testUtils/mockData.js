export const testSections = 
    [
        {
            id: "0",
            slug: "section1",
            sectionTitle: "Section 1",
            active: true
        },
        {
            id: "1",
            slug: "section2",
            sectionTitle: "Section 2",
            active: true
        }
    ]

export const testProducts = 
    [
        {
            id: "0",
            image1: {fields: {file: {url: "1"}}},
            image2: {fields: {file: {url: "2"}}},
            image3: {fields: {file: {url: "3"}}},
            price: 5,
            title: "Product 1",
            description: "Description",
            productSection: {fields: {slug: "section1"}},
            active: true
        },
        {
            id: "1",
            image1: {fields: {file: {url: "1"}}},
            image2: {fields: {file: {url: "2"}}},
            image3: {fields: {file: {url: "3"}}},
            price: 5,
            title: "Product 2",
            description: "Description",
            productSection: {fields: {slug: "section1"}},
            active: true
        },
        {
            id: "2",
            image1: {fields: {file: {url: "1"}}},
            image2: {fields: {file: {url: "2"}}},
            image3: {fields: {file: {url: "3"}}},
            price: 5,
            title: "Product 3",
            description: "Description",
            productSection: {fields: {slug: "section2"}},
            active: true
        }
    ]

export const testProduct = 
    {
        id: 0,
        image1: {fields: {file: {url: "1"}}},
        image2: {fields: {file: {url: "2"}}},
        image3: {fields: {file: {url: "3"}}},
        price: 5,
        title: "Product 1",
        description: "Description",
        productSection: {fields: {slug: "section1"}},
        active: true
    }

export const testCartProduct = 
[{
    id: 0,
    image1: {fields: {file: {url: "1"}}},
    image2: {fields: {file: {url: "2"}}},
    image3: {fields: {file: {url: "3"}}},
    price: 5,
    title: "Product 1",
    description: "Description",
    productSection: {fields: {slug: "section1"}},
    active: true,
    quantity: 1
}]
