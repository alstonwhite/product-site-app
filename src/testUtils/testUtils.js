import React from "react";
import { Router, Route } from "react-router-dom";
import { render } from "@testing-library/react";
import { createMemoryHistory } from "history";

export const renderWithRouter = (ui, 
    {
        path="/",
        route="/",
        history = createMemoryHistory({ initialEntries: [route] }) 
    }={}) => ( 
    render( 
        <Router history={history}> 
            <Route path={path} children={ui} /> 
        </Router> 
    ) 
)