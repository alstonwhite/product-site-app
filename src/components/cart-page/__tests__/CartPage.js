import React from 'react'
import { cleanup, fireEvent, screen } from '@testing-library/react'
import CartPage from '../CartPage'
import { testProduct } from '../../../testUtils/mockData'
import { renderWithRouter } from '../../../testUtils/testUtils'


afterEach(cleanup)

jest.mock('react-redux', () => {
    return {
    ...jest.requireActual('react-redux'),
    // cart initializes with item 'testCartProduct' -- hardcoded bc import out of scope 
    useSelector: () => 
    [{
        id: 0,
        image1: {fields: {file: {url: "1"}}},
        image2: {fields: {file: {url: "2"}}},
        image3: {fields: {file: {url: "3"}}},
        price: 5,
        title: "Product 1",
        description: "Description",
        productSection: {fields: {slug: "section1"}},
        active: true,
        quantity: 1
    }],
    useDispatch: jest.fn()
    }
});

test('renders proper cart tiles', () => {
    renderWithRouter(<CartPage/>, {path: "/cart", route: "/cart"})
    
    expect(screen.getAllByTestId('cart-tile')).toHaveLength(1)
})

// mocked checkout call not triggering
test('calls checkout when button clicked', () => {
    const mockCheckout = jest.fn();
    jest.mock('../../../utils/checkoutStripe', () => ({
        ...jest.requireActual('../../../utils/checkoutStripe'),
        checkout: () => mockCheckout
    }));
    renderWithRouter(<CartPage/>, {path: "/cart", route: "/cart"})
  
    fireEvent.click(screen.getByText('Checkout'))
  
    expect(mockCheckout).toHaveBeenCalledTimes(1)
})

test('calculates and displays cart value', () => {
  const mockCheckout = jest.fn();
  jest.mock('../../../utils/checkoutStripe', () => ({
      ...jest.requireActual('../../../utils/checkoutStripe'),
      checkout: () => mockCheckout
  }));
  renderWithRouter(<CartPage/>, {path: "/cart", route: "/cart"})

  expect(screen.getByText(testProduct.price)).toBeInTheDocument()
})

test('snapshot', () => {
  const {container} = renderWithRouter(<CartPage/>, {path: "/cart", route: "/cart"})
  expect(container.firstChild).toMatchSnapshot()
})