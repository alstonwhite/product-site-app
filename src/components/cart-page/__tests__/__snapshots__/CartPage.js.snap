// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<div
  class="cart-page"
  id="cart-page"
>
  <h1>
    Cart
  </h1>
  <div
    class="cart-grid__items"
  >
    <div
      class="cart-tile"
      data-testid="cart-tile"
      id="cart-tile"
    >
      <a
        class="cart-tile__content"
        href="/product/0"
      >
        <img
          alt="logo"
          class="cart-tile__content__img"
          src="1"
        />
        <div
          class="cart-tile__content__name"
        >
          Product 1
        </div>
        <div
          class="cart-tile__content__price"
        >
          <span>
            $
          </span>
          5
        </div>
      </a>
      <div
        class="cart-tile__action"
      >
        <input
          class="cart-tile__action__qty"
          min="0"
          step="1"
          type="number"
          value="1"
        />
        <button
          class="cart-tile__action__remove btn"
        >
          X
        </button>
      </div>
    </div>
  </div>
  <div
    class="cart-page__footer"
  >
    <div
      class="cart-page__footer__value"
    >
      Total: $5
    </div>
    <button
      class="cart-page__footer__checkout btn"
    >
      Checkout
    </button>
  </div>
</div>
`;
